### DISCUSSION TITLE
Demo：NebulaGraph 的 Graph RAG - siwei.io

### DISCUSSION THREAD

**Discussion by giscus** [siwei-io/siwei-io.github.io#26] [Comments]
# Demo：NebulaGraph 的 Graph RAG - siwei.io

https://www.siwei.io/demos/graph-rag/

**Comment by shaunT0ny** [comment]
demo程序挂了🥲
> RuntimeError: The services status exception: [services: ('****), status: BAD]

**Reply by wey-gu** [reply]
抱歉，现在恢复啦！

**Comment by bg-vc** [comment]
请问，图数据库换成DGraph，是否可以实现类似的效果？

**Reply by wey-gu** [reply]
自己手撸理论上是也可行的哈，未来不过我们还会做更多深度集成的事情，不一定都能做到哈~~
要不要试试 NebulaGraph？ http://nebula-up.siwei.io/ 可以一键服务器部署试玩  https://hub.docker.com/extensions/weygu/nebulagraph-dd-ext 可以一键桌面试玩

**Comment by DreamingJ** [comment]
程序似乎又挂了
``` APIRemovedInV1: You tried to access openai.Completion, but this is no longer supported in openai>=1.0.0 - see the README at https://github.com/openai/openai-python for the API. You can run `openai migrate` to automatically upgrade your codebase to use the 1.0.0 interface. Alternatively, you can pin your installation to the old version, e.g. `pip install openai==0.28` A detailed migration guide is available here: https://github.com/openai/openai-python/discussions/742 ```

**Reply by wey-gu** [reply]
啊，我应该 pin version 的，openai dev day 之后 api break change 了，感谢告知！

**Reply by wey-gu** [reply]
recovered

**Reply by DreamingJ** [reply]
现在还是异常 
`RuntimeError: The services status exception: [services: ('***', 9669), status: BAD]`

**Reply by wey-gu** [reply]
Sorry, 另一个 failure point，恢复了哈。

**Comment by duanzhichen** [comment]
openai.api_type = "azure"
openai.api_base = "https://<foo-bar>.openai.azure.com"
openai.api_version = "2022-12-01"
os.environ["OPENAI_API_KEY"] = "youcannottellanyone"
openai.api_key = os.getenv("OPENAI_API_KEY")

llm = AzureOpenAI(
    engine="<foo-bar-deployment>",
    temperature=0,
    openai_api_version=openai.api_version,
    model_kwargs={
        "api_key": openai.api_key,
        "api_base": openai.api_base,
        "api_type": openai.api_type,
        "api_version": openai.api_version,
    },
)

我想用本公司开发的大模型接口，上面的代码怎么改造？

**Reply by wey-gu** [reply]
是兼容 openai api 的？用 llama index 的 openai: https://docs.llamaindex.ai/en/stable/examples/llm/openai.html 给 base 和 key

**Reply by duanzhichen** [reply]
我问的是百度或阿里有没有类似AzureOpenAI的安装包？

**Reply by duanzhichen** [reply]
openai.api_type = "azure"
openai.api_base = "https://.openai.azure.com"
openai.api_version = "2022-12-01"
os.environ["OPENAI_API_KEY"] = "youcannottellanyone"

更具体点：这几个参数值能不能替换成文心一言、阿里通义千问或其它国内大模型？

**Reply by wey-gu** [reply]
是可以的，两种方法：1，他们有 llamaindex llms 的实现 2，他们有 openai api 兼容，前者参考 https://www.siwei.io/demo-dumps/local-llm/Graph_RAG_Local.html ，后者用我上边提到的方式

**Reply by wey-gu** [reply]
比如，llamaindex 支持所有 langchain 的 llms 实现，参考 https://docs.llamaindex.ai/en/stable/examples/llm/langchain.html

**Reply by duanzhichen** [reply]
我的vx: tingyuxuan891192，方便加个微信深入交流一下llama_index和知识图谱的集合吗？

**Reply by duanzhichen** [reply]
endpoint_url = "http://127.0.0.1:8000" # LLM API
embedding_model = "shibing624/text2vec-base-chinese"

import logging
import sys

logging.basicConfig(stream=sys.stdout, level=logging.INFO) # logging.DEBUG for more verbose output
# logging.getLogger().addHandler(logging.StreamHandler(stream=sys.stdout))


# LLM
from langchain.llms import ChatGLM
from langchain import PromptTemplate, LLMChain


llm = ChatGLM(
    endpoint_url=endpoint_url,
    max_token=2048,
    top_p=0.9,
    temperature=1,
    model_kwargs={
        "sample_model_args": False,
    },
)

llm.with_history = False


运行这段代码一直报“ConnectionRefusedError:[WinError 10061]  由于目标计算机积极决绝，无法连接。”的错误，是需要配置密钥和选择的模型吗？

**Comment by Sun0308** [comment]
您好，好像第二个demo又不行了，报错：openai.error.InvalidRequestError: This app has encountered an error. The original error message is redacted to prevent data leaks. Full error details have been recorded in the logs (if you're on Streamlit Cloud, click on 'Manage app' in the lower right of your app).

**Reply by wey-gu** [reply]
感谢提醒，我找时间修一下哈~

**Reply by wey-gu** [reply]
修复了哈，是因为 Azure Ada model 的版本更新 :) 感谢

**Comment by yjh-Coder** [comment]
请问这个问题一般是为什么？
Exception ignored in: <function NebulaGraphStore.__del__ at 0x00000206B03EE430>
Traceback (most recent call last):
  File "E:\Env\anaconda\envs\ChatGLM-6B\lib\site-packages\llama_index\graph_stores\nebulagraph.py", line 248, in __del__
  File "E:\Env\anaconda\envs\ChatGLM-6B\lib\site-packages\nebula3\gclient\net\SessionPool.py", line 273, in close
  File "E:\Env\anaconda\envs\ChatGLM-6B\lib\site-packages\nebula3\gclient\net\Session.py", line 288, in _sign_out
  File "E:\Env\anaconda\envs\ChatGLM-6B\lib\site-packages\nebula3\gclient\net\Connection.py", line 205, in signout
  File "E:\Env\anaconda\envs\ChatGLM-6B\lib\site-packages\nebula3\graph\GraphService.py", line 1603, in signout
  File "E:\Env\anaconda\envs\ChatGLM-6B\lib\site-packages\nebula3\graph\GraphService.py", line 1606, in send_signout
AttributeError: 'NoneType' object has no attribute 'CALL'

**Reply by wey-gu** [reply]
可以稳定复现么？我一下子也不知道为啥

**Reply by yjh-Coder** [reply]
您好，我是在https://www.siwei.io/demo-dumps/local-llm/Graph_RAG_Local.html ，尝试复现您的实验，这个错误是连接NebulaGraph 时，没有手动关闭连接，报错，后续我添加了一个手动close就没问题，但是我在实验中没有复现出文档这么好的效果，我使用的就是文档中推荐的 ChatGLM2-6B，是存在什么优化吗？

**Comment by AiX-im** [comment]
非常感谢您的分享，Graph RAG的结合很惊艳。请问一下在知识图谱的检索与在向量数据库中检索的开销占比如何呢？开销是否与提取出实体的数量与top K 的取值有关，大概的关系如何呢，哪个检索的开销更大，谢谢

**Reply by wey-gu** [reply]
两个数据库本身的 retrieve 的开销我还真没比较过，不过我理解 graph 更接近数据库的模式，vector db 上的 approximate nearest neighbors 的 computing overhead 更大一些。

不过一般来说两者不是比较的，而是协作、协同、互补的哈。

**Reply by ShilongYuan** [reply]
https://zhuanlan.zhihu.com/p/660722382 有人总结了优劣，但是很疑惑，我原以为kg会比vector快的

**Reply by wey-gu** [reply]
这是 WenQi 的文章的翻译、节选（这个文章我有帮忙 review、给出建议），原文在：https://betterprogramming.pub/7-query-strategies-for-navigating-knowledge-graphs-with-llamaindex-ed551863d416?gi=8b3e84780f5a 这个文章只是比较了当时在 llamaIndex 里的一些状况哈，我后边又有很大的进展哈

**Reply by wey-gu** [reply]
> https://zhuanlan.zhihu.com/p/660722382 有人总结了优劣，但是很疑惑，我原以为kg会比vector快的

不是很容易说快慢的，有很多复杂的原因哈，我在 llamaindex 当时的实现里有很多地方可以优化的，绝对说 latency 的话 kg 没有更慢（不过 graph rag 里我在里边的实现中间有多 llm shots 的场景）

现在我这边探索 SOTA 的方法是 chain of exploration 还没有开源，效果好很多很多，敬请关注

**Comment by wey-gu** [comment]
@AiX-im 你好哈，好像你用邮件回复，消息没有记录在这个 thread 里。

> 非常期待您的新工作，感谢您的开源精神。我们其实原本不做llm的工作，而是Graph方面的，有一些system优化的底子，希望能follow到您的脚步。

感谢你的 nice word! 你的 https://github.com/iDC-NEU/HyTGraph 工作非常非常酷，不知道你是否感兴趣来我们团队实习哈？

**Reply by AiX-im** [reply]
感谢您的邀请哈，目前还在苦苦的干博士课题（也有合作的企业），遗憾没法跟大佬面对面学习了。

**Reply by wey-gu** [reply]
> 感谢您的邀请哈，目前还在苦苦的干博士课题（也有合作的企业），遗憾没法跟大佬面对面学习了。

let us know when you are free to!

**Comment by BBBeatrice** [comment]
您好，demo程序运行之后又报错了，可以麻烦看下吗
RuntimeError: The services status exception: [services: ('***', 9669), status: BAD]

**Comment by snowwalkk** [comment]
你好，Vector vs Graph + Vector的那个demo，报错openai.error.InvalidRequestError: This app has encountered an error. The original error message is redacted to prevent data leaks. Full error details have been recorded in the logs (if you're on Streamlit Cloud, click on 'Manage app' in the lower right of your app).
